torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -0,0 +1,163 @@
+#include <torch/nn/modules/transformer.h>
+#include <torch/nn/init.h>
+
+#include <torch/types.h>
+#include <torch/utils.h>
+
+#include <cmath>
+#include <cstdint>
+
+namespace F = torch::nn::functional;
+
+namespace torch {
+namespace nn {
+
+  // ========================TransformerDecoderLayerImpl=========================
+  TransformerDecoderLayerImpl::TransformerDecoderLayerImpl(
+    const TransformerDecoderLayerOptions& options_ )
+    : options(options_){
+    reset();
+  }
+
+  void TransformerDecoderLayerImpl::reset() {
+    ///initialize self attention
+    self_attn = register_module(
+      "self_attn",
+      MultiheadAttention(
+        MultiheadAttentionOptions(options.d_model(), options.nhead())
+        .dropout(options.dropout())));
+
+    ///initialize Dropout, post self attention
+    dropout1 = register_module("dropout1",
+                               Dropout(DropoutOptions().p(options.dropout())));
+
+    ///initialize Normalization, post self attention
+    norm1 = register_module(
+      "norm1",
+      LayerNorm(LayerNormOptions(std::vector<int64_t> {options.d_model()})));
+
+    ///initialize multihed attention
+    multihead_attn = register_module(
+      "multihead_attn",
+      MultiheadAttention(
+        MultiheadAttentionOptions(options.d_model(), options.nhead())
+        .dropout(options.dropout())));
+
+    ///initialize post multi-headed attention dropout layer
+    dropout2 = register_module(
+      "dropout2", Dropout(DropoutOptions().p(options.dropout())));
+
+   ///initialize post multi-headed attention Normalization
+    norm2 = register_module(
+      "norm2", LayerNorm(
+        LayerNormOptions(std::vector<int64_t> {options.d_model()})));
+
+    ///Initialize Feed forward first linear layer
+    linear1 = register_module(
+      "linear1",
+      Linear(LinearOptions(options.d_model(), options.dim_feedforward())));
+
+    ///initialize Feed forward dropout layer
+    dropout = register_module(
+      "dropout",
+      Dropout(DropoutOptions().p(options.dropout())));
+
+    ///initialize Feed forward second linear layer
+    linear2 = register_module(
+      "linear2",
+      Linear(LinearOptions(options.dim_feedforward(), options.d_model())));
+
+    ///initialize dropout, post feed forward
+    dropout3 = register_module(
+      "dropout3",
+      Dropout(DropoutOptions().p(options.dropout())));
+
+    ///initialize normalization, post feed forward
+    norm3 = register_module(
+      "norm3",
+      LayerNorm(LayerNormOptions(std::vector<int64_t> {options.d_model()})));
+  }
+
+  void TransformerDecoderLayerImpl::reset_parameters() {
+    self_attn->reset();
```

Comment:
Please see my change for encoder layer here:
https://github.com/pytorch/pytorch/pull/42633/files#diff-0705518cdb48c453dd00e335768e4db7R35
You only need to call reset_parameters() for each submodule if any.

Note that dropout module has no reset_parameter().